### ISSUE TITLE
IMailFolder.Count() on large folder (32k emails) takes a very very long time

### ISSUE THREAD

**Issue by AJCheeseman** [jstedfast/MailKit#1444]
I opened my "Processed" folder using IMAP and called the Count() command on the IMailFolder object and it has been running for 10 minutes and still hasn't come back with a number yet.
I am trying to get the count, but I was trying to get the UniqueId of an email I just MoveTo'ed into that folder. 
In some searches, I think I wonder if I should have used this instead.
`folder.Fetch(1,-1,MessageSummaryItems.UniqueId)`

**Platform (please complete the following information):**
 - OS: Windows
 - .NET Framework: .NET 4.5
 - MailKit Version: 2.15.0

**Exception**
Not really an exception, just stuck waiting for the result to come back. Here is a copy from the Threads window
```
System.dll!System.Net.Sockets.Socket.Receive
[External Code]
MailKit.dll!MailKit.Net.NetworkStream.Read(byte[] buffer, int offset, int count) Line 165
[External Code]
MailKit.dll!MailKit.Net.Imap.ImapStream.ReadAheadAsync(int atleast, bool doAsync, System.Threading.CancellationToken cancellationToken) Line 316
MailKit.dll!MailKit.Net.Imap.ImapStream.ReadTokenAsync(string specials, bool doAsync, System.Threading.CancellationToken cancellationToken) Line 698
MailKit.dll!MailKit.Net.Imap.ImapEngine.ReadTokenAsync(bool doAsync, System.Threading.CancellationToken cancellationToken) Line 849
MailKit.dll!MailKit.Net.Imap.ImapCommand.StepAsync(bool doAsync) Line 840
MailKit.dll!MailKit.Net.Imap.ImapEngine.IterateAsync(bool doAsync) Line 2124
MailKit.dll!MailKit.Net.Imap.ImapEngine.RunAsync(MailKit.Net.Imap.ImapCommand ic, bool doAsync) Line 2174
MailKit.dll!MailKit.Net.Imap.ImapFolder.GetMessageAsync(int index, bool doAsync, System.Threading.CancellationToken cancellationToken, MailKit.ITransferProgress progress) Line 4773
MailKit.dll!MailKit.Net.Imap.ImapFolder.GetMessage(int index, System.Threading.CancellationToken cancellationToken, MailKit.ITransferProgress progress) Line 4831
MailKit.dll!MailKit.Net.Imap.ImapFolder.GetEnumerator() Line 5979
[External Code]
Evaluation of: folder.Count()
```

**To Reproduce**
1. Open Folder with a lot of emails in it
2. call folder.Count() 
3. ....

**Expected behavior**

I expected it to return the folder count of 35,552

**Code Snippets**
```csharp
using (ImapClient client = new ImapClient())
{
	await Task.Run(() => EstablishImapConnection(this, client));

	if (UniqueId.TryParse(ei.UniqueID.ToString(), out UniqueId uid))
	{
		var folder = client.Inbox;
		folder.Open(FolderAccess.ReadWrite);
		var ProcessedFolder = client.GetFolder("INBOX/Processed Emails");
		folder.MoveTo(uid, ProcessedFolder);
		folder.Close();
		folder = ProcessedFolder;
		folder.Open(FolderAccess.ReadOnly);
		ei.UniqueID = (int)folder.Fetch(new List<int> { _folder.Count()_ - 1 }, MessageSummaryItems.UniqueId).First().UniqueId.Id;
	} else
	{
		throw new InvalidOperationException("Unable to Parse UniqueID");
	}
	await client.DisconnectAsync(true);
}//using (ImapClient client = new ImapClient())
```

**Protocol Logs**
I only included the start of the Protocol Log because reading the log I'm now noticing a few things (see additional context)
> Connected to imaps://outlook.office365.com:993/
> S: * OK The Microsoft Exchange IMAP4 service is ready. [-----------------------------]
> C: B00000000 CAPABILITY
> S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS ID UNSELECT CHILDREN IDLE NAMESPACE LITERAL+
> S: B00000000 OK CAPABILITY completed.
> C: B00000001 AUTHENTICATE -----------------------------------------------------------
> S: B00000001 OK AUTHENTICATE completed.
> C: B00000002 CAPABILITY
> S: * CAPABILITY IMAP4 IMAP4rev1 AUTH=PLAIN AUTH=XOAUTH2 SASL-IR UIDPLUS MOVE ID UNSELECT CLIENTACCESSRULES CLIENTNETWORKPRESENCELOCATION BACKENDAUTHENTICATE CHILDREN IDLE NAMESPACE LITERAL+
> S: B00000002 OK CAPABILITY completed.
> C: B00000003 NAMESPACE
> S: * NAMESPACE (("" "/")) NIL NIL
> S: B00000003 OK NAMESPACE completed.
> C: B00000004 LIST "" "INBOX"
> S: * LIST (\Marked \HasChildren) "/" INBOX
> S: B00000004 OK LIST completed.
> C: B00000005 SELECT INBOX
> S: * 1 EXISTS
> S: * 0 RECENT
> S: * FLAGS (\Seen \Answered \Flagged \Deleted \Draft $MDNSent)
> S: * OK [PERMANENTFLAGS (\Seen \Answered \Flagged \Deleted \Draft $MDNSent)] Permanent flags
> S: * OK [UIDVALIDITY 14] UIDVALIDITY value
> S: * OK [UIDNEXT 348418] The next unique identifier value
> S: B00000005 OK [READ-WRITE] SELECT completed.
> C: B00000006 LIST "" "INBOX/Processed Emails"
> S: * LIST (\Marked \HasChildren) "/" "INBOX/Processed Emails"
> S: B00000006 OK LIST completed.
> C: B00000007 UID MOVE 348417 "INBOX/Processed Emails"
> S: * OK [COPYUID 417 348417 199704]
> S: * 1 EXPUNGE
> S: * 0 EXISTS
> S: B00000007 OK MOVE completed.
> C: B00000008 UNSELECT
> S: B00000008 OK UNSELECT completed.
> C: B00000009 EXAMINE "INBOX/Processed Emails"
> S: * 35552 EXISTS
> S: * 1 RECENT
> S: * FLAGS (\Seen \Answered \Flagged \Deleted \Draft $MDNSent)
> S: * OK [PERMANENTFLAGS ()] Permanent flags
> S: * OK [UIDVALIDITY 417] UIDVALIDITY value
> S: * OK [UIDNEXT 199705] The next unique identifier value
> S: B00000009 OK [READ-ONLY] EXAMINE completed.
> C: B00000010 FETCH 1 (BODY.PEEK[])
> S: * 1 FETCH (BODY[] {8643}

**Additional context**
I noticed a few things in the protocol log

1. The process is peeking every email, assume this is to get the REAL count because the number could have changed since the EXISTS value was returned when the folder was opened
2. I had glossed over the fact that when sending the UniqueId to the MoveTo command, it would return to me the new UniqueId in the destination folder (this wasn't a thing in the last Imap library I used before switching to MailKit) but when I saw the COPYUID had another UID value I clued in.

***In Summary*** 
**I've solved my issue** of needing the new UID after a move **by saving the function's return value**, but I am still submitting this issue because I believe it is a bug that IMailFolder.Count() must peek each message to get the folder count (I really hope I've just implemented MailKit in the wrong way resulting in this behavior)

**Comment by jstedfast** [comment]
The `IMailFolder` interface doesn't have a `Count()` method (`Count()` is a LINQ extension method for types that implement the `IEnumerable` interface).

You should have used the `Count` *property* instead of the LINQ extension method.

The LINQ extension method (which exists in System.Linq, not MailKit) fetches every single message in the folder in order to get a count of messages.

Marking as `invalid` because of that - but if MailKit itself did this, then yea, I would consider it a bug :-)

I regret ever making `IMailFolder` implement `IEnumerable<MimeMessage>` because of this, but I can't undo that without breaking some projects that actually exist out there that make use of that.
